### PR TITLE
Refactor stat collector

### PR DESCRIPTION
**Pull request by khushbr**
**Is your feature request related to a problem? Please provide an existing Issue # , or describe.**
`StatsCollector` is used by RCA-Agent and PerformanceAnalyzer Plugin to publish service metrics. We run this collector through `ScheduledMetricCollectorsExecutor` and the executor can mark the collector as SLOW and subsequently MUTED if the collector is slow. A muted collector doesn't emit any metrics and thus we can end up missing the helpful service metrics.

This PR contains changes to skip muting for  `StatsCollector` and re-factoring to move away from `StatExceptionCode` to Metric Aggregators. 

Adding more granular code change details:
* PerformanceAnalyzerApp.java : Removing Default Exception code. (Moving to Aggregator metric framework, does not require default codes)
* PerformanceAnalyzerThreads.java : Moved Thread failure exception codes to `ReaderMetrics`. Each time a thread dies, the respective metric will be emitted via the catch block in `ThreadProvider#createThreadForRunnable()`
* ScheduledMetricCollectorsExecutor.java : `StatsCollector` is now non-mutable. This is required to ensure that we always have service metrics available to debug issues.
* StatExceptionCode.java : Moved exception codes to relevant Aggregators. 
* StatsCollector.java :  Removed `logMetric(final String metricName)` and `logException()`. 
* FlowUnitRxTask.java : Removing duplicate metric `VERTEX_BUFFER_FULL_METRIC`. We emit `RCA_VERTEX_RX_BUFFER_FULL_ERROR` in ReceivedFlowUnitStore.
* ReaderMetricsProcessor.java : Emitting `READER_METRICS_EMIT_TIME` instead of using older `TIMING_STATS` and `STATS_DATA` data structure.

**Describe the solution you are proposing**

**Describe alternatives you've considered**

**Additional context**

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/performance-analyzer-rca/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by sruti1312** [comment]
Branch conflicts needs to be resolved.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/ScheduledMetricCollectorsExecutor.java [hunk 1]
```diff
@@ -120,8 +122,8 @@ public void run() {
                         PerformanceAnalyzerMetricsCollector collector = entry.getKey();
                         if (collector.getState()
                                 == PerformanceAnalyzerMetricsCollector.State.MUTED) {
-                            StatsCollector.instance()
-                                    .logMetric(StatExceptionCode.COLLECTORS_MUTED.toString());
+                            PerformanceAnalyzerApp.WRITER_METRICS_AGGREGATOR.updateStat(
```
Comment (sruti1312) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/ScheduledMetricCollectorsExecutor.java:125]: Is there a way to include which collector was muted?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/StatsCollector.java [hunk 2]
```diff
@@ -82,24 +97,58 @@ public static StatsCollector instance() {
         return statsCollector;
     }
 
+    @Override
+    public void collectMetrics(long startTime) {
+        Map<String, AtomicInteger> currentCounters = counters;
+        counters = new ConcurrentHashMap<>();
+
+        // currentCounters.putIfAbsent(StatExceptionCode.TOTAL_ERROR.toString(), new
+        // AtomicInteger(0));
```
Comment (sruti1312) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/StatsCollector.java:106]: Do we need this? If not can we remove it?

File: src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/RcaVerticesMetrics.java [hunk 3]
```diff
@@ -36,6 +36,15 @@
  * metrics added by each RCA vertex in RCA graph. All metrics under this category are RCA specific
  */
 public enum RcaVerticesMetrics implements MeasurementSet {
+    INVALID_OLD_GEN_SIZE("InvalidOldGenSize", "count", Collections.singletonList(Statistics.COUNT)),
+    OLD_GEN_RECLAMATION_INEFFECTIVE(
+            "OldGenReclamationIneffective", "count", Collections.singletonList(Statistics.COUNT)),
+    OLD_GEN_CONTENDED("OldGenContended", "count", Collections.singletonList(Statistics.COUNT)),
+    OLD_GEN_OVER_OCCUPIED(
+            "OldGenOverOccupied", "count", Collections.singletonList(Statistics.COUNT)),
```
Comment (sruti1312) [inline src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/RcaVerticesMetrics.java:44]: nit: Can we have a space between each one (just to be uniform)?

File: src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/ReaderMetrics.java [hunk 4]
```diff
@@ -106,7 +154,14 @@
...
                     Statistics.MEAN,
                     Statistics.COUNT,
-                    Statistics.SUM));
+                    Statistics.SUM)),
+
+    /**
+     * A blanket exception code for {@link
+     * org.opensearch.performanceanalyzer.reader.ReaderMetricsProcessor} failures.
+     */
+    OTHER("Other", "count", Collections.singletonList(Statistics.COUNT));
```
Comment (sruti1312) [inline src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/ReaderMetrics.java:163]: Can we name this OTHER_READER_METRICS_PROCESSOR_ERRORS or something like this? Or atleast OTHER_ERRORS? Not sure how much info we will get from ```OTHER``` when we see them in prod.

File: src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/WriterMetrics.java [hunk 5]
```diff
@@ -34,6 +34,42 @@
...
+    /** Thread ID is no loner exists */
+    JVM_THREAD_ID_NO_LONGER_EXISTS("JVMThreadIdNoLongerExists"),
+
+    /** Tracks the number of muted collectors */
+    COLLECTORS_MUTED("CollectorsMutedCount"),
+
+    METRICS_REMOVE_ERROR("MetricsRemoveError"),
+    MASTER_METRICS_ERROR("MasterMetricsError"),
+    MASTER_NODE_NOT_UP("MasterNodeNotUp"),
+    OPENSEARCH_REQUEST_INTERCEPTOR_ERROR("OpenSearchRequestInterceptorError"),
```
Comment (sruti1312) [inline src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/WriterMetrics.java:70]: If possible can have a comment on what these metrics denote?

**Comment by khushbr** [comment]
Closing this PR and opening a new one from forked repo https://github.com/opensearch-project/performance-analyzer-rca/pull/37

File: src/main/java/org/opensearch/performanceanalyzer/rca/net/tasks/FlowUnitRxTask.java [hunk 6]
```diff
@@ -76,9 +74,7 @@ public void run() {
             LOG.warn(
                     "Dropped a flow unit because the vertex buffer was full for vertex: {}",
                     vertex);
-            StatsCollector.instance().logMetric(RcaConsts.VERTEX_BUFFER_FULL_METRIC);
```
Comment (sruti1312) [inline src/main/java/org/opensearch/performanceanalyzer/rca/net/tasks/FlowUnitRxTask.java:79]: Do we want emit equivalent metric here with the other framework?

File: src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java [hunk 7]
```diff
@@ -57,10 +52,10 @@ public void testNullThreadInfo() {
 
         ThreadList.parseAllThreadInfos(infos);
 
-        Map<String, AtomicInteger> counters = StatsCollector.instance().getCounters();
+        /*Map<String, AtomicInteger> counters = StatsCollector.instance().getCounters();
```
Comment (sruti1312) [inline src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java:55]: Can we remove this if we are not using this anymore?

File: src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java [hunk 8]
```diff
@@ -57,10 +52,10 @@ public void testNullThreadInfo() {
...
 
         Assert.assertEquals(
-                counters.get(StatExceptionCode.JVM_THREAD_ID_NO_LONGER_EXISTS.toString()).get(), 1);
+                counters.get(PerformanceAnalyzerApp.WRITER_METRICS_AGGREGATOR.getValues(WriterMetrics.JVM_THREAD_ID_NO_LONGER_EXISTS).toString()).get(), 1);*/
```
Comment (sruti1312) [inline src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java:58]: Do we want to uncomment this?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/StatsCollector.java (hunk 2 above)
Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/StatsCollector.java:106]: Done.

File: src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/RcaVerticesMetrics.java (hunk 3 above)
Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/RcaVerticesMetrics.java:44]: Ack, Done.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/ScheduledMetricCollectorsExecutor.java (hunk 1 above)
Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/ScheduledMetricCollectorsExecutor.java:125]: Good point, converted to `NAMED_COUNTERS`

File: src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/WriterMetrics.java (hunk 5 above)
Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/WriterMetrics.java:70]: Done.

File: src/main/java/org/opensearch/performanceanalyzer/rca/net/tasks/FlowUnitRxTask.java (hunk 6 above)
Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/rca/net/tasks/FlowUnitRxTask.java:79]: So, we have a repetitive metric here. We already emit `RCA_VERTEX_RX_BUFFER_FULL_ERROR` inside [receivedFlowUnitStore.enqueue()](https://github.com/opensearch-project/performance-analyzer-rca/blob/main/src/main/java/org/opensearch/performanceanalyzer/rca/net/ReceivedFlowUnitStore.java#L84)

File: src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java (hunk 7 above)
Comment (khushbr) [reply src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java:55]: Done, fixed this test.

File: src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java (hunk 8 above)
Comment (khushbr) [reply src/test/java/org/opensearch/performanceanalyzer/jvm/ThreadListTest.java:58]: Done, fixed this test.

File: src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/ReaderMetrics.java (hunk 4 above)
Comment (khushbr) [reply src/main/java/org/opensearch/performanceanalyzer/rca/framework/metrics/ReaderMetrics.java:163]: `ReaderMetricsProcessor` is one of the places this is used, and this is sort of meant as a blanket exception code. Will keep it like this for now and we can re-visit.

**Pull request closed without merging**